apps/bfd-model/bfd-model-rda/README.md

```diff
@@ -17,3 +17,155 @@ uses Spring to build up its JPA configuration and Spring does not have the same
+|---------------------|---------------------|-----------------------|--------------------------|
+| mcs_adjustments     | mcs_adjustments     | McsAdjustment         | mcs_adjustment.proto     |
+| mcs_audits          | mcs_audits          | McsAudit              | mcs_audit.proto          |
+| mcs_details         | mcs_details         | McsDetail             | mcs_detail.proto         |
+| mcs_diagnosis_codes | mcs_diagnosis_codes | McsDiagnosisCode      | mcs_diagnosis_code.proto |
+| mcs_locations       | mcs_locations       | McsLocation           | mcs_location.proto       |
+
+This diagram shows all of the MCS related tables along with their primary keys.
+
+```
+                               ┌─────────────────────┐
+                               │ mcs_adjustments     │
+                          ┌────┤ - idr_clm_hd_icn    │
+                          │    │ - priority          │
+                          │    └─────────────────────┘
+                          │
+                          │    ┌─────────────────────┐
+                          │    │ mcs_audits          │
+                          ├────┤ - idr_clm_hd_icn    │
+                          │    │ - priority          │
+                          │    └─────────────────────┘
+                          │
+┌──────────────────┐      │    ┌─────────────────────┐
+│ mcs_claims       │      │    │ mcs_details         │
+│ - idr_clm_hd_icn ◄──────┼────┤ - idr_clm_hd_icn    │
+└──────────────────┘      │    │ - priority          │
+                          │    └─────────────────────┘
+                          │
+                          │    ┌─────────────────────┐
+                          │    │ mcs_diagnosis_codes │
+                          ├────┤ - idr_clm_hd_icn    │
+                          │    │ - priority          │
+                          │    └─────────────────────┘
+                          │
+                          │    ┌─────────────────────┐
+                          │    │ mcs_locations       │
+                          └────┤ - idr_clm_hd_icn    │
+                               │ - priority          │
+                               └─────────────────────┘
+```
+
+## RDA API Progress
+
+The RDA ETL pipeline jobs consume data by calling a gRPC service in the RDA API.  This service
+returns a stream of claim messages.  Each message represents either a new claim or an update
+to an existing claim.  Every message includes a sequence number.  These sequence numbers
+can be used to track progress within the stream and to resume from a specific location
+in the stream in a later RPC call.
+
+The `rda_api_progress` table is used to store the last processed sequence number for each claim type.
+The pipeline jobs update the table as they process incoming claims.  The jobs can be forced to
+replay data by stopping the job, changing the `last_sequence_number` value for the appropriate
+claim type, and restarting the pipeline job.
```

Comment:
Is this different from the way non-RDA claims are processed, or am I just misssing information about how the pipeline works? (For instance, we can replay non-RDA loads by running in idempotent mode without touching a database directly)

Reply:
The RDA pipeline is always idempotent in the sense that reingesting the same claims multiple times will not really change any claim data except last updated timestamps.  But the progress table is only consulted when the pipeline jobs start and then updated continuously as they ingest claims.  As a consequence if you change the value while the job is running you would just see your change overwritten without it doing what you want.

The alternative would be to query the current value from the progress table before updating it (which happens millions of times per run) prior to each update and maybe abort the process if the value is different than expected.  We did not implement it that way since reingestion is expected to be extremely rare in practice (like only in response to a data consistency incident or something) and it would be somewhat slower to do it that way.

Reply:
I see, so RDA has a more granular record of what's been loaded, in which case it's unlikely we'd need to rerun the entire load anyway. Any error/failure would only need to resume at the last progress update, so its not like the non-RDA pipeline. Thanks for explaining!